### PR TITLE
Rate limiting non-staff users.

### PR DESCRIPTION
**Pull request by basisbit**
### Before you begin

* I understand my contributions may be rejected for any reason
* I understand my contributions are for the benefit of this imageboard software
* I understand my contributions are licensed under the GNU AGPLv3

- [x] I understand all of the above

---

Cherry-picked three commits regarding web-ui rate-limiting of non-staff users.

Co-authored by liamwhite

### CODE REVIEW

File: lib/philomena_web/plugs/limit_plug.ex [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        |> Controller.put_flash(:error, error)
+        |> Controller.redirect(external: conn.assigns.referrer)
+        |> Conn.halt()
+    end
+  end
+
+  defp is_staff(%User{role: "admin"}), do: true
+  defp is_staff(%User{role: "moderator"}), do: true
+  defp is_staff(%User{role: "assistant"}), do: true
+  defp is_staff(_), do: false
```
Comment (alamais-df) [inline lib/philomena_web/plugs/limit_plug.ex:63]: I feel like the staff bypass should canonically be done via `can?` and ability.ex, but it's not a strong feeling.

Comment (basisbit) [reply lib/philomena_web/plugs/limit_plug.ex:63]: would incremental improvement (with further PRs, whenever someone worked on this) be okay for you?

Comment (alamais-df) [reply lib/philomena_web/plugs/limit_plug.ex:63]: It'll probably fall into a black hole and never be addressed, but that's fine.  Like I said, not a strong feeling.  :v

Comment (basisbit) [reply lib/philomena_web/plugs/limit_plug.ex:63]: Would be nice to actually implement error/debug logging into the app. Those both tasks could be resolve at the same time:
- change those limits into configuration options,
- make those error messages adjustable depending on configuration options,
- offer a simple logging API
- change hardcoded limits, permission-checks, and similar into easily adjustable options.

**Pull request merged**